### ISSUE TITLE
line in comments is not used in Code.quoted_to_algebra/2

### ISSUE THREAD

**Issue by Eiji7** [elixir-lang/elixir#11001]
Hi, I'm too excited with latest changes and just can't wait for `release candidate` for next `Elixir` version. :smiling_imp: 

### Environment

* Elixir & Erlang/OTP versions (elixir --version): 
```elixir
Erlang/OTP 23 [erts-11.2.2] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [hipe]

Elixir 1.13.0-dev (c6145d0) (compiled with Erlang/OTP 23)
```
* Operating system: 
```bash
$ uname -a
Linux archlinux 5.12.1-arch1-1 #1 SMP PREEMPT Sun, 02 May 2021 12:43:58 +0000 x86_64 GNU/Linux

$ cat /etc/os-release
NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://www.archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://bugs.archlinux.org/"
LOGO=archlinux
```

### Current behavior

`line` in `comments` is not (properly) used in `Code.quoted_to_algebra/2`, see:
```elixir
iex> source = """
defmodule Example do
  # something
  def sample, do: :ok
end
"""

iex> {:ok, quoted, comments} = Code.string_to_quoted_with_comments source
{:ok,
 {:defmodule, [line: 1],
  [
    {:__aliases__, [line: 1], [:Example]},
    [do: {:def, [line: 3], [{:sample, [line: 3], nil}, [do: :ok]]}]
  ]},
 [%{line: 2, next_eol_count: 1, previous_eol_count: 1, text: "# something"}]}

iex> algebra = Code.quoted_to_algebra quoted, comments: comments
{:doc_force,
 {:doc_cons, {:doc_group, "# something", :self},
  {:doc_cons, :doc_line,
   {:doc_group,
    {:doc_cons, {:doc_string, "defmodule", 9},
     {:doc_force,
      {:doc_cons,
       {:doc_cons,
        {:doc_group,
         {:doc_cons, {:doc_cons, " ", {:doc_nest, "Example", :cursor, :break}},
          " do"}, :self},
        {:doc_nest,
         {:doc_cons, :doc_nil,
          {:doc_cons, :doc_line,
           {:doc_group,
            {:doc_group,
             {:doc_cons, {:doc_string, "def", 3},
              {:doc_force,
               {:doc_cons,
                {:doc_cons,
                 {:doc_group,
                  {:doc_cons, {:doc_cons, " ", {:doc_nest, {...}, ...}}, " do"},
                  :self},
                 {:doc_nest,
                  {:doc_cons, :doc_nil, {:doc_cons, :doc_line, {...}}}, 2,
                  :always}}, {:doc_cons, :doc_line, "end"}}}}, :self}, :self}}},
         2, :always}}, {:doc_cons, :doc_line, "end"}}}}, :self}}}}

iex> algebra |> Inspect.Algebra.format(80) |> IO.puts()
# something
defmodule Example do
  def sample do
    :ok
  end
end
```

### Expected behavior

The output should be as same as from `source` variable.

**Comment by josevalim** [comment]
Pass the `:token_metadata` option when calling string_to_quoted, otherwise we don't have the proper do/end placements, which leads us to place comments incorrectly. I will improve the docs.

**Comment by Eiji7** [comment]
Oh, my bad then. I was thinking that it would "just work". I read documentation carefully and now see why those options are important. Many thanks!

**Comment by Eiji7** [comment]
@josevalim A small update … I tried same code with `opts` suggested by documentation:

```elixir
opts = [
  literal_encoder: &{:ok, {:__block__, &2, [&1]}},
  token_metadata: true,
  unescape: true
]

source = """
defmodule Example do
  # something
  def sample, do: :ok
end
"""

{:ok, quoted, comments} = Code.string_to_quoted_with_comments source, opts

algebra = Code.quoted_to_algebra quoted, comments: comments

algebra |> Inspect.Algebra.format(80) |> IO.puts()
```

and the result is:
```elixir
defmodule Example do
  # something
  def sample,
      [do: :ok]
end
```

It's really close to original, but still `do` keyword in function declaration is different. Am I doing something wrong?

**Comment by josevalim** [comment]
Looks like a bug. @doorgan, can you please take a look? Note I have slightly changed the tests structure in master, but we can probably add a new case to `formatter_ast_test.exs`.

**Comment by josevalim** [comment]
Btw, two-element tuples are also literals... it may be we are double wrapping them in blocks.

**Comment by doorgan** [comment]
Yes, I will take a look into it

**Comment by doorgan** [comment]
The issue is that I'm looking for a `{:do, _}` for kw blocks, but I'm not also looking for `{:__block__, _, [:do]}` which is what we receive if the AST is already normalized.

I also found the same issue if brackets are omitted in expressions like `@type a when b: :c`, they would be converted to `@type a when [b: :c]`

It should be fixed with #11002

**Comment by doorgan** [comment]
Also, if you use the `unescape: false` option in `Code.string_to_quoted_with_comments/2`, you may need to pass the `escape: false` option to `Code.quoted_to_algebra/2`, otherwise escape sequences like `\n` will be escaped twice, resulting in `\\\\n` instead of `\\n`
